### PR TITLE
Adding big block matrices

### PR DESCRIPTION
**Pull request by CohenCyril**
##### Motivation for this change

This is part of #207.  We add the analoguous of  `\matrix_(i, j) _` `\row_i _`, `\col_i _` and `diag_mx` but for block matrices instead of single elements.

##### Things done/to do

- [x] added corresponding entries in `CHANGELOG_UNRELEASED.md` (do not edit former entries)
- [x] added corresponding documentation in the headers

##### Automatic note to reviewers

Read [this Checklist](https://github.com/math-comp/math-comp/wiki/Checklist-for-following,-reviewing-and-playing-with-a-PR#checklist-for-reviewing-a-pr) and make sure there is a milestone.

### CODE REVIEW

File: CHANGELOG_UNRELEASED.md [hunk 1]
```diff
@@ -26,6 +26,52 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 - in `interval.v`, new lemmas: `ge_pinfty`, `le_ninfty`, `gt_pinfty`, `lt_ninfty`.
 
+- in `order.v`
```
Comment (thery) [inline CHANGELOG_UNRELEASED.md:29]: Could this be a separate commit?

File: mathcomp/algebra/matrix.v [hunk 2]
```diff
@@ -49,6 +49,21 @@ From mathcomp Require Import div prime binomial ssralg finalg zmodp countalg.
 (*                   same width).            \ Ad /                           *)
 (* block_mx Aul Aur Adl Adr == the block matrix / Aul Aur \                   *)
 (*                                              \ Adl Adr /                   *)
+(* \mxrow_(i < n) B i ==  the block m * (\sum_i p_i) matrix                   *)
```
Comment (thery) [inline mathcomp/algebra/matrix.v:52]: Are they places where this new constructor could be used in the current library?

File: mathcomp/ssreflect/order.v [hunk 3]
```diff
@@ -399,6 +403,20 @@ From mathcomp Require Import path fintype tuple bigop finset div prime.
...
+(* whenever the type is porderType, and their monotonicity is provided if     *)
+(* this order is total. The theory is in the module Order (Order.enum_valK,   *)
+(* Order.enum_rank_inK, etc) but Order.Enum can be imported to shorten these. *)
+(******************************************************************************)
+(* We provide an opaque monotonous bijection tagnat.val / tagnat.rank between *)
+(* the finite types {i : 'I_n & 'I_(p_ i)} and 'I_(\sum_i p_ i):              *)
+(*  tagnat.val   : 'I_(\sum_i p_ i) -> {i : 'I_n & 'I_(p_ i)}                 *)
+(*  tagnat.rank  : {i : 'I_n & 'I_(p_ i)} -> 'I_(\sum_i p_ i)                 *)
+(*  tagnat.val1  : 'I_(\sum_i p_ i) -> 'I_n                                   *)
+(*  tagnat.val2  : forall p : 'I_(\sum_i p_ i), 'I_(p_ (tagnat.val1 p))       *)
```
Comment (thery) [inline mathcomp/ssreflect/order.v:417]: I was expected `tagnat.tag` for` val1`  and `tagnat.nat` for `val2` but maybe it is not a good idea.

File: mathcomp/algebra/matrix.v [hunk 4]
```diff
@@ -3686,3 +3716,67 @@ End mxOver.
...
+  \matrix_(j, k)
+   if tagnat.val1 j =P tagnat.val1 k isn't ReflectT e then 0
+   else B_ (tagnat.val1 k)
+     (cast_ord (congr1 p_ e) (tagnat.val2 j)) (tagnat.val2 k).
+
+Notation "\mxcol_ i E" := (mxcol (fun i => E)) : ring_scope.
+Notation "\mxrow_ i E" := (mxrow (fun i => E)) : ring_scope.
+Notation "\mxdiag_ i E" := (mxdiag (fun i => E)) : ring_scope.
+
+Lemma mulmx_row_col m p
```
Comment (thery) [inline mathcomp/algebra/matrix.v:3739]: A pity these new constructions come with very little theorems associated with
them.
I was thinking for example about the trace of `mxdiag` or the transpose of `mxcol`

File: CHANGELOG_UNRELEASED.md (hunk 1 above)
Comment (CohenCyril) [reply CHANGELOG_UNRELEASED.md:29]: How do you suggest to split? (They are related in cascade, each idem depends on the previous one)

File: mathcomp/algebra/matrix.v (hunk 2 above)
Comment (CohenCyril) [reply mathcomp/algebra/matrix.v:52]: Not yet, but that's all part of a bigger plan: https://github.com/math-comp/math-comp/pull/207/files#diff-f7a28e39a52f7013e677ad2bab9930222f5af8c0df4acb10e978bf2a1a2fa37fR291-R318

File: mathcomp/ssreflect/order.v (hunk 3 above)
Comment (CohenCyril) [reply mathcomp/ssreflect/order.v:417]: I'm really open to any name here. Yours are funny, it also made me think `tagnat.val1` might not have to be a natural number... so that there could be a further generalisation... as long as it is an ordered finite type...

File: mathcomp/algebra/matrix.v (hunk 4 above)
Comment (CohenCyril) [reply mathcomp/algebra/matrix.v:3739]: Sure. I did not do the closure, I wrote the bare minimum for https://github.com/math-comp/math-comp/pull/207/files#diff-f7a28e39a52f7013e677ad2bab9930222f5af8c0df4acb10e978bf2a1a2fa37fR291-R318 to go through...
So what should I add?
- the trace of mxdiag
- the transpose of all of them
- I guess `mulmx_row_diag`
- maybe `mulmx_col_row` if it makes sense...
other ideas?

Comment (CohenCyril) [reply mathcomp/algebra/matrix.v:3739]: Maybe stuff with `0`s and `1`s and `a%:M`

File: mathcomp/ssreflect/order.v (hunk 3 above)
Comment (CohenCyril) [reply mathcomp/ssreflect/order.v:417]: Actually for now (because of the restriction on finite ordered types) its preferable to stick to the first component being a nat. It's a bit silly, but it could be generalized later. Also I prefer to keep `val1` and `val2` string-related to `val`, because it's easier for me to remember...

File: mathcomp/algebra/matrix.v (hunk 4 above)
Comment (thery) [reply mathcomp/algebra/matrix.v:3739]: I guess there are lots in fact, for example the rank of `\mx_diag` also is there an equivalent of `submx`

File: CHANGELOG_UNRELEASED.md (hunk 1 above)
Comment (thery) [reply CHANGELOG_UNRELEASED.md:29]: I was just thinking that the first 3 items were mostly independent of block matrices. 
Tagnat instead seems more closely related to matrices.

File: mathcomp/algebra/matrix.v (hunk 2 above)
Comment (thery) [reply mathcomp/algebra/matrix.v:52]: I was thinking at places where one has been using \matrix_ where one of your constructor could be more natural but a rapid grep showed me nothing

File: mathcomp/ssreflect/order.v (hunk 3 above)
Comment (thery) [reply mathcomp/ssreflect/order.v:417]: Ok just it made explicit the encoding of tagnat but that'fine

**Comment by CohenCyril** [comment]
@thery  I did the split -> #710

**Comment by CohenCyril** [comment]
I also dramatically enhanced the block matrix theory... It will be here soon.

**Comment by CohenCyril** [comment]
@thery I ended up adding a couple lemmas, feels rather complete now.

File: mathcomp/algebra/matrix.v (hunk 2 above)
Comment (CohenCyril) [reply mathcomp/algebra/matrix.v]: Thanks for checking

**Comment by CohenCyril** [comment]
@thery FYI I added a commit renaming `val` to `sig`: it causes less undesirable name-clashes, especially if one wants to import `tagnat`, as in [matrix.v](https://github.com/math-comp/math-comp/pull/703/files#diff-f457cb0bf11677ea475ce7d07aa9d8ce79d060ef642432808d92ceb55d6a3fc5R3743).

**Comment by thery** [comment]
@cyril looks good. Will I have a closer look on monday.

File: CHANGELOG_UNRELEASED.md [hunk 5]
```diff
@@ -52,6 +52,62 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
...
+    the components of the former type via the functions `sig1`, `sig2`
+    and `Rank`. We provide the following lemmas on these definitions:
+    `card`, `sigK`, `sig_inj`, `rankK`, `rank_inj`, `sigE12`,
+    `rankE`, `sig2K`, `Rank1K`, `Rank2K`, `rank_bij`, `sig_bij `,
+    `rank_bij_on`, `sig_bij_on`, `le_sig`, `le_sig1`, `le_rank`,
+    `le_Rank`, `lt_sig`, `lt_rank`, `lt_Rank`, `eq_Rank`, `rankEsum`,
+    `RankEsum`, `rect`, and `eqRank`.
+
+- in `matrix.v`, seven new definitions:
+  + `mxblock` `mxcol`, `mxrow` and `mxdiag` with notations
```
Comment (thery) [inline CHANGELOG_UNRELEASED.md:68]: mxblock,

File: CHANGELOG_UNRELEASED.md [hunk 6]
```diff
@@ -52,6 +52,62 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
...
+    `\mxblock_(i < m, j < n) B_ i j`, `\mxcol_(i < m) B_ i`,
+    `\mxrow_(j < n) B_ j`, and `\mxdiag_(i < n) B_ i` (and variants
+    without explicit `< n`), to form big matrices described by blocks.
+  + `submxblock`, `submxcol` and `submxrow` to extract blocks from the
```
Comment (thery) [inline CHANGELOG_UNRELEASED.md:72]: why not submxdiag?

File: CHANGELOG_UNRELEASED.md [hunk 7]
```diff
@@ -52,6 +52,62 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
...
+    former constructions.
+  + We provides the following lemmas on these definitions:
```
Comment (thery) [inline CHANGELOG_UNRELEASED.md:74]: provide

**Review by thery — COMMENTED** [review]
impressive!

File: CHANGELOG_UNRELEASED.md (hunk 6 above)
Comment (CohenCyril) [reply CHANGELOG_UNRELEASED.md:72]: This would be redundant with `submxblock A i i`

Comment (CohenCyril) [reply CHANGELOG_UNRELEASED.md:72]: I could add it as a notation, but I'm not convinced it is very useful

File: CHANGELOG_UNRELEASED.md (hunk 5 above)
Comment (CohenCyril) [reply CHANGELOG_UNRELEASED.md:68]:
```suggestion
  + `mxblock`, `mxcol`, `mxrow` and `mxdiag` with notations
```

File: CHANGELOG_UNRELEASED.md (hunk 7 above)
Comment (CohenCyril) [reply CHANGELOG_UNRELEASED.md:74]:
```suggestion
  + We provide the following lemmas on these definitions:
```

File: CHANGELOG_UNRELEASED.md (hunk 6 above)
Comment (CohenCyril) [reply CHANGELOG_UNRELEASED.md:72]: cf https://github.com/math-comp/math-comp/pull/703/files#diff-f457cb0bf11677ea475ce7d07aa9d8ce79d060ef642432808d92ceb55d6a3fc5R4371-R4373

Comment (thery) [reply CHANGELOG_UNRELEASED.md:72]: maybe just adding a comment

Comment (CohenCyril) [reply CHANGELOG_UNRELEASED.md:72]: Where and what comment would you think is appropriate?
Something like ``there is no analogous for `mxdiag` because one can simply apply `submxblock A i i` ``?

File: CHANGELOG_UNRELEASED.md (hunk 7 above)
Comment (CohenCyril) [reply CHANGELOG_UNRELEASED.md:74]: @thery if you resolve the issue by hand I cannot commit the "suggested change" any more. Please leave this kind of comment open until github automatically resolves them when I push the change. :smile_cat:

File: CHANGELOG_UNRELEASED.md (hunk 5 above)
Comment (CohenCyril) [reply CHANGELOG_UNRELEASED.md]: done

File: CHANGELOG_UNRELEASED.md (hunk 7 above)
Comment (CohenCyril) [reply CHANGELOG_UNRELEASED.md]: done

File: CHANGELOG_UNRELEASED.md [hunk 8]
```diff
@@ -52,6 +52,62 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
...
+    `le_Rank`, `lt_sig`, `lt_rank`, `lt_Rank`, `eq_Rank`, `rankEsum`,
+    `RankEsum`, `rect`, and `eqRank`.
+
+- in `matrix.v`, seven new definitions:
+  + `mxblock`, `mxcol`, `mxrow` and `mxdiag` with notations
+    `\mxblock_(i < m, j < n) B_ i j`, `\mxcol_(i < m) B_ i`,
+    `\mxrow_(j < n) B_ j`, and `\mxdiag_(i < n) B_ i` (and variants
+    without explicit `< n`), to form big matrices described by blocks.
+  + `submxblock`, `submxcol` and `submxrow` to extract blocks from the
+    former constructions.
```
Comment (CohenCyril) [inline CHANGELOG_UNRELEASED.md:73]:
```suggestion
    former constructions. There is no analogous for `mxdiag` because
    one can simply apply `submxblock A i i`.
```

File: CHANGELOG_UNRELEASED.md (hunk 6 above)
Comment (thery) [reply CHANGELOG_UNRELEASED.md:72]: yes in the documention of `matrix.v`.
This makes me realised that there is no doc for `subblock` and the others.

Comment (CohenCyril) [reply CHANGELOG_UNRELEASED.md:72]: See the new revision

Comment (thery) [reply CHANGELOG_UNRELEASED.md:72]: this should go in the header of `matrix.v` with the documentation of `subblock` and al

Comment (CohenCyril) [reply CHANGELOG_UNRELEASED.md:72]: I added a comment about the absence of `submxdiag` juste below the line you pointed out.
For the other subblocks I am tempted to open an issue instead of fixing it here.

Comment (thery) [reply CHANGELOG_UNRELEASED.md:72]: I am not sure I undertand. 
You have introduced new constructors `submxblock` ... so you should add their documentation in the header of `matrix`. I think the comment about the absense of `submxdiag` should go there too.

Comment (CohenCyril) [reply CHANGELOG_UNRELEASED.md:72]: Ah you are right, sorry I got confused between the changelog and the header

**Comment by CohenCyril** [comment]
@thery I addressed all your comments.

**Review by thery — APPROVED** [review]

**Comment by CohenCyril** [comment]
@thery Thanks!
